File: python/MCIntro_program1.py
import sys
import time

def simple_rand(seed):
    seed = (seed * 1664525 + 1013904223) & 0xFFFFFFFFFFFFFFFF
    rand_val = ((seed >> 16) & 0x7FFF) / 32768.0
    return rand_val, seed

def func(x):
    return 4.0 / (1.0 + x * x)

if len(sys.argv) != 2:
    print(f"Usage: {sys.argv[0]} N")
    sys.exit(1)

try:
    n = int(sys.argv[1])
except ValueError:
    print("Invalid number")
    sys.exit(1)

seed = int(time.time())
sum_fx = 0.0
sum_fx2 = 0.0

for _ in range(n):
    x, seed = simple_rand(seed)
    fx = func(x)
    sum_fx += fx
    sum_fx2 += fx * fx

mean = sum_fx / n
mean2 = sum_fx2 / n
variance = mean2 - mean * mean

print(f"Integral = {mean}, variance = {variance}")

File: rust/src/MCIntro_program1.rs
use std::env;
use std::process;
use std::time::{SystemTime, UNIX_EPOCH};

fn simple_rand(seed: &mut u64) -> f64 {
    *seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
    ((*seed >> 16) & 0x7FFF) as f64 / 32768.0
}

fn func(x: f64) -> f64 {
    4.0 / (1.0 + x * x)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} N", args[0]);
        process::exit(1);
    }
    let n: u64 = args[1].parse().expect("Invalid number");
    let mut sum = 0.0;
    let mut sum2 = 0.0;
    let start = SystemTime::now();
    let mut seed = start.duration_since(UNIX_EPOCH).unwrap().as_secs();
    for _ in 0..n {
        let x = simple_rand(&mut seed);
        let fx = func(x);
        sum += fx;
        sum2 += fx * fx;
    }
    let mean = sum / n as f64;
    let mean2 = sum2 / n as f64;
    let variance = mean2 - mean * mean;
    println!("Integral = {}, variance = {}", mean, variance);
}